### PR TITLE
Cleanup exited containers, fixes #2

### LINKED ISSUE #2: Can't delete some images
Using a stateless deis 1.12.3, some images can't be deleted because it's in use. The containers should be removed first.

```
Error response from daemon: Conflict, cannot delete 9a654fb551d6 because the container 11379f004d9f is using it, use -f to force
Error: failed to remove images: [10.xx.yy.zz:5000/production:v45]
```

The command used was:

```
docker run \
 -v /var/run/docker.sock:/var/run/docker.sock \
 -v /var/lib/docker:/var/lib/docker \
 --rm \
 twobox/deis-cleanup
```

### PR DESCRIPTION
**Pull request by szymonpk**
Script by default will cleanup all exited containers, it requires at
least docker 1.8.3 due to "--format" deis images option.

### CODE REVIEW

**Comment by szymonpk** [comment]
@st @smt116 take a look.

File: deis-cleanup.rb [hunk 1]
```diff
@@ -30,11 +31,30 @@
   puts output
 end
 
+remove_containers = lambda do |container|
```
Comment (smt116) [inline deis-cleanup.rb]: is it `remove_containers` or `remove_container`?

File: deis-cleanup.rb [hunk 2]
```diff
@@ -30,11 +31,30 @@
...
+  command = "docker rm #{container[:id]}"
+  command.prepend "echo #{container[:name]}: " if DRY_RUN
+  output, _status = Open3.capture2e(command)
+  puts output
+end
```
Comment (smt116) [inline deis-cleanup.rb:36]: This is exactly the same code as in lines 27-32 except of the docker subcommand

File: deis-cleanup.rb [hunk 3]
```diff
@@ -30,11 +31,30 @@
...
+
+output, status = Open3.capture2e("docker ps -a --format '{{.ID}};{{.Names}}' --filter 'exited=0'")
+abort output unless status.success?
+
+rows = CSV.parse(output, col_sep: ";")
+containers = rows.map { |row| %i(id name).zip(row).to_h }
```
Comment (smt116) [inline deis-cleanup.rb]: I would put that before lambdas definitions just after images assignment.

File: deis-cleanup.rb [hunk 4]
```diff
@@ -30,11 +31,30 @@
...
+
 puts "deis-cleanup started."
 
+if CLEANUP_CONTAINERS
+  puts "removing containers"
+  containers.reject { |container| container[:name].match(EXCLUDE_REGEX) }.each(&remove_containers)
+end
```
Comment (smt116) [inline deis-cleanup.rb:52]: oh wait.. do we even need to fetch exited containers when `CLEANUP_CONTAINERS` is set to false?

Comment (smt116) [inline deis-cleanup.rb:52]: It could be chained to the `map` from line 45 if not

File: deis-cleanup.rb [hunk 5]
```diff
@@ -6,35 +6,57 @@
...
-# - EXCLUDE_REGEX - regex filter for repository name (default: `alpine|deis|blackhole|none|datadog|cleanup|heroku|python`).
+# - EXCLUDE_REGEX - regex filter for repository name (default: `alpine|deis|blackhole|none|datadog|cleanup|heroku|python).
 
 require "open3"
 require "csv"
 require "yaml"
 
+def execute(command, abort_on_failure = false)
+  output, status = Open3.capture2e(command)
+  abort output unless abort_on_failure && status.success?
```
Comment (smt116) [inline deis-cleanup.rb]: abort on success?

File: deis-cleanup.rb [hunk 6]
```diff
@@ -6,35 +6,57 @@
...
+
+remove_container = lambda do |container|
+  command = "docker rm #{container[:id]}"
+  command.prepend "echo #{container[:name]}: " if DRY_RUN
+  execute(command)
+end
 
-output, status = Open3.capture2e("docker images")
-abort output unless status.success?
+output = execute("docker images", true)
```
Comment (smt116) [inline deis-cleanup.rb]: remove this empty line

File: deis-cleanup.rb [hunk 7]
```diff
@@ -6,35 +6,57 @@
...
-  puts output
-end
-
 puts "deis-cleanup started."
 
+if CLEANUP_CONTAINERS
+  puts "removing containers"
+
+  output = execute("docker ps -a --format '{{.ID}};{{.Names}}' --filter 'exited=0'", true)
+
```
Comment (smt116) [inline deis-cleanup.rb]: remove empty line

**Comment by smt116** [comment]
👍

**Pull request merged**